iml-api/src/graphql/mod.rs

```diff
@@ -485,6 +464,99 @@ impl QueryRoot {
 
         Ok(xs)
     }
+
+    #[graphql(arguments(
+        limit(description = "optional paging limit, defaults to all rows"),
+        offset(description = "Offset into items, defaults to 0"),
+        dir(description = "Sort direction, defaults to asc"),
+        message(description = "Pattern to search for in message"),
+        fqdn(description = "Pattern to search for in FQDN"),
+        tag(description = "Pattern to search for in tag"),
+        start_datetime(description = "Start of the time period of logs"),
+        end_datetime(description = "End of the time period of logs"),
+        message_class(description = "Array of log message classes"),
+        severity(description = "Upper bound of log severity"),
+    ))]
+    async fn logs(
+        context: &Context,
+        limit: Option<i32>,
+        offset: Option<i32>,
+        dir: Option<SortDir>,
+        message: Option<String>,
+        fqdn: Option<String>,
+        tag: Option<String>,
+        start_datetime: Option<String>,
+        end_datetime: Option<String>,
+        message_class: Option<Vec<MessageClass>>,
+        severity: Option<LogSeverity>,
+    ) -> juniper::FieldResult<LogResponse> {
+        let dir = dir.unwrap_or_default();
+
+        let start_datetime = start_datetime
+            .map(|s| s.parse::<chrono::DateTime<Utc>>())
+            .map_or(Ok(None), |r| r.map(Some))?;
+
+        let end_datetime = end_datetime
+            .map(|s| s.parse::<chrono::DateTime<Utc>>())
+            .map_or(Ok(None), |r| r.map(Some))?;
+
+        let message_class: Vec<_> = message_class
+            .filter(|v| !v.is_empty())
+            .unwrap_or_else(|| vec![MessageClass::Normal])
+            .into_iter()
+            .map(|i| i as i16)
+            .collect();
+
+        let severity = severity.unwrap_or(LogSeverity::Debug) as i16;
+
+        let results = sqlx::query_as!(
+            LogMessageRecord,
+            r#"
+                    SELECT * FROM chroma_core_logmessage t
+                    WHERE t.message LIKE $4
+                      AND t.fqdn LIKE $5
+                      AND t.tag LIKE $6
+                      AND ($7::TIMESTAMPTZ IS NULL OR t.datetime >= $7)
+                      AND ($8::TIMESTAMPTZ IS NULL OR t.datetime < $8)
+                      AND ARRAY[t.message_class] <@ $9
+                      AND t.severity <= $10
+                    ORDER BY
+                        CASE WHEN $3 = 'ASC' THEN t.datetime END ASC,
+                        CASE WHEN $3 = 'DESC' THEN t.datetime END DESC
+                    OFFSET $1 LIMIT $2"#,
+            offset.unwrap_or(0) as i64,
+            limit.map(|x| x as i64),
+            dir.deref(),
+            message.unwrap_or_else(|| "%".into()),
+            fqdn.unwrap_or_else(|| "%".into()),
+            tag.unwrap_or_else(|| "%".into()),
```

Comment:
AFAIK `LIKE '%'` is not optimized out.

Reply:
Any suggestions?

Reply:
You're right, I've implemented a check to remove filter when it's not needed